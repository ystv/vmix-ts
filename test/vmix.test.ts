import { EventEmitter } from "node:events";
import * as fs from "node:fs/promises";
import * as path from "path";
import VMixConnection from "../src/vmix";

class MockSocket extends EventEmitter {
  write = jest.fn(
    (payload: unknown, encoding: string, cb: (err?: Error) => void) =>
      cb(undefined),
  );
  setEncoding = jest.fn();
}

jest.mock("node:net", () => ({
  connect: () => {
    const sock = new MockSocket();
    process.nextTick(() => {
      sock.emit("connect");
    });
    return sock;
  },
}));

const nextTick = () => new Promise((resolve) => process.nextTick(resolve));

describe("VMixConnection", () => {
  let vmix: VMixConnection;
  let sock: MockSocket;
  beforeEach(async () => {
    vmix = await VMixConnection.connect();
    sock = vmix["sock"] as unknown as MockSocket;
  });
  describe("send/receive", () => {
    test("request and response", async () => {
      // vmix["send"] lets us access vmix.send even though it's private
      const res = vmix["send"]("FUNCTION", "test");
      expect(sock.write).toHaveBeenCalledWith(
        "FUNCTION test\r\n",
        "utf-8",
        expect.any(Function),
      );
      sock.emit("data", "FUNCTION OK test\r\n");
      expect(res).resolves.toEqual(["test", ""]);
    });
    test("no-arg request", async () => {
      vmix["send"]("XML");
      expect(sock.write).toHaveBeenCalledWith(
        "XML\r\n",
        "utf-8",
        expect.any(Function),
      );
    });
    test("binary response", async () => {
      const res = vmix["send"]("FUNCTION", "test");
      sock.emit("data", "FUNCTION 5\r\nhello");
      expect(res).resolves.toEqual(["", "hello"]);
    });
    test("binary response across multiple chunks", async () => {
      const res = vmix["send"]("FUNCTION", "test");
      sock.emit("data", "FUNCTION 5\r\n");
      sock.emit("data", "hello");
      expect(res).resolves.toEqual(["", "hello"]);
    });
    test("response split across chunks", async () => {
      const res = vmix["send"]("FUNCTION", "test");
      sock.emit("data", "FUNCTION");
      sock.emit("data", " OK\r\n");
      expect(res).resolves.toEqual(["", ""]);
    });
    test("message and binary", async () => {
      const res = vmix["send"]("FUNCTION", "test");
      sock.emit(
        "data",
        "FUNCTION 28 This is a message in addition to the binary data\r\n" +
          "This is optional binary data",
      );
      expect(res).resolves.toEqual([
        "This is a message in addition to the binary data",
        "This is optional binary data",
      ]);
    });
    test("error handling", async () => {
      const res = vmix["send"]("FUNCTION", "test");
      sock.emit("data", "FUNCTION ER test\r\n");
      expect(res).rejects.toThrow("test");
    });
    test("two requests", async () => {
      const r1 = vmix["send"]("FUNCTION", "test1");
      const r2 = vmix["send"]("FUNCTION", "test2");
      await nextTick();
      expect(sock.write).toHaveBeenCalledWith(
        "FUNCTION test1\r\n",
        "utf-8",
        expect.any(Function),
      );
      sock.emit("data", "FUNCTION OK test1\r\n");
      expect(r1).resolves.toEqual(["test1", ""]);
      await nextTick();
      expect(sock.write).toHaveBeenCalledWith(
        "FUNCTION test2\r\n",
        "utf-8",
        expect.any(Function),
      );
      sock.emit("data", "FUNCTION OK test2\r\n");
      expect(r2).resolves.toEqual(["test2", ""]);
    });
  });

  test("getFullState", async () => {
    const testXML = await fs.readFile(
      path.join(__dirname, "__testdata__", "vmix.xml"),
      { encoding: "utf-8" },
    );
    const res = vmix.getFullState();
    expect(sock.write).toHaveBeenCalledWith(
      "XML\r\n",
      "utf-8",
      expect.any(Function),
    );
    sock.emit("data", `XML ${testXML.length}\r\n${testXML}`);
    expect(res).resolves.toMatchSnapshot();
  });
});
