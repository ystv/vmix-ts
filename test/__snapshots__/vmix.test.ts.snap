// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VMixConnection getFullState 1`] = `
{
  "edition": "4K",
  "inputs": [
    {
      "duration": 0,
      "key": "418f6d70-ec7e-464a-8abc-92280776f5a0",
      "loop": false,
      "number": 1,
      "position": 0,
      "shortTitle": "TV Mix",
      "state": "Paused",
      "title": "TV Mix",
      "type": "Mix",
    },
    {
      "duration": 0,
      "key": "1c077daf-caf6-4371-be56-ca11736d12dd",
      "loop": false,
      "number": 2,
      "position": 0,
      "shortTitle": "Transparent",
      "state": "Paused",
      "title": "Transparent",
      "type": "Colour",
    },
    {
      "duration": 279573,
      "items": [
        {
          "selected": true,
          "source": "C:\\media\\Test 0.mp4",
        },
        {
          "selected": false,
          "source": "C:\\media\\Test 1.mp4",
        },
        {
          "selected": false,
          "source": "C:\\media\\Test 22.mp4",
        },
        {
          "selected": false,
          "source": "C:\\media\\Test 3.mov",
        },
        {
          "selected": false,
          "source": "C:\\media\\Test 4.mp4",
        },
      ],
      "key": "705aa6bc-632e-4714-8759-0966703c154f",
      "loop": false,
      "number": 3,
      "position": 0,
      "selectedIndex": 0,
      "shortTitle": "VTs",
      "state": "Paused",
      "title": "VTs - Test 0.mp4",
      "type": "VideoList",
    },
    {
      "audioBusses": "M",
      "balance": 0,
      "duration": 5000,
      "key": "410bc785-3d4b-4d0d-96a0-58e0ed78edde",
      "loop": false,
      "muted": true,
      "number": 4,
      "position": 0,
      "shortTitle": "YY_Transition_1.mov",
      "solo": false,
      "state": "Paused",
      "title": "YY_Transition_1.mov",
      "type": "Video",
      "volume": 100,
    },
    {
      "duration": 0,
      "key": "254bf3d2-6b74-4dc3-a46f-bac16b5c4fee",
      "loop": false,
      "number": 5,
      "position": 0,
      "shortTitle": "Colour",
      "state": "Paused",
      "title": "Colour",
      "type": "Colour",
    },
    {
      "duration": 0,
      "items": [
        {
          "selected": true,
          "source": "C:\\media\\Test.png",
        },
      ],
      "key": "be25c595-60e1-49b1-873b-1c1f7fde77b4",
      "loop": false,
      "number": 6,
      "position": 0,
      "selectedIndex": 0,
      "shortTitle": "Stills",
      "state": "Paused",
      "title": "Stills - Test.png",
      "type": "VideoList",
    },
    {
      "duration": 0,
      "key": "ba632fc4-90b6-42db-afd0-fbc73487fa23",
      "loop": false,
      "number": 7,
      "position": 0,
      "shortTitle": "Test.png",
      "state": "Paused",
      "title": "Test.png",
      "type": "Image",
    },
    {
      "audioBusses": "M",
      "balance": 0,
      "duration": 60899,
      "key": "b5076d09-361c-4185-8086-9c45c9de75de",
      "loop": false,
      "muted": true,
      "number": 8,
      "position": 0,
      "shortTitle": "ES_Young Mystery Detectives - Trailer Worx.wav",
      "solo": false,
      "state": "Paused",
      "title": "ES_Young Mystery Detectives - Trailer Worx.wav",
      "type": "AudioFile",
      "volume": 100,
    },
  ],
  "preset": "C:\\Test.vmix",
  "version": "26.0.0.44",
}
`;
